File: examples/matlab/mfiles/e_HierarchicalVectorInterp.m
%% Hierarchical vector interpolation

clear ; close all ; path=pwd();
[s,e]=regexp(path,'.+(?=[\\/]{1,2}examples)');
path=path(s:e); addpath(genpath(path)); clear s e path
%% 1) Definition of the interpolation tree
% From the leaves to the root.

dimInput = 2;
dimOutput = 2;

defNode11.Domain = Domain(3);
f111v = VertexFunction(@(b) b.^2, @(b) [2*b(1,1,:),zeros(1,1,size(b,3));zeros(1,1,size(b,3)),2*b(2,1,:)], "f111v", dimInput, dimOutput);
f112v = VertexFunction(@(b) b.^3, @(b)  [3*b(1,1,:).^2,zeros(1,1,size(b,3));zeros(1,1,size(b,3)),3*b(2,1,:).^2], "f112v", dimInput, dimOutput);
f113v = VertexFunction(@(b) 0.1*b, @(b) ones(size(b)).*eye(2)*0.1, "f113v", dimInput, dimOutput);
defNode11.Children = [f111v,f112v,f113v];
defNode11.Label = "n11"; % Neveu convention (any unique labeling is possible)
defNode11.Penalization = Penalization("simp",2);
Node11 = Interpolation(defNode11);

defNode12.Domain = Domain(2);

f121v = VertexFunction(@(b) ones(size(b)), @(b) zeros(2,2,size(b,3)), "f121v", dimInput, dimOutput); % constant
f122v = VertexFunction(@(b) mult([1,2;3,4],b), @(b) [1,2;3,4].*ones(1,1,size(b,3)), "f122v", dimInput, dimOutput);
defNode12.Children = [f121v,f122v];
defNode12.Label = "n12";
defNode12.Penalization = Penalization("simp",2);
Node12 = Interpolation(defNode12);

defNode1.Domain = Domain(2);
defNode1.Children = [Node11,Node12];
defNode1.Label = "n1";
defNode1.Penalization = Penalization("simp",2);
Node1 = Interpolation(defNode1);


% singleton domains to keep homogeneous children of Node 0
defNode2.Domain = Domain(1);
defNode2.Children = VertexFunction(@(b) 2*b.^3, @(b)  2*[3*b(1,1,:).^2,zeros(1,1,size(b,3));zeros(1,1,size(b,3)),3*b(2,1,:).^2], "f21v", dimInput, dimOutput);
defNode2.Label = "n2";
Node2 = Interpolation(defNode2);

defNode3.Domain = Domain(1);
defNode3.Children =  VertexFunction(@(b) mult([1,2;8,4],b), @(b) [1,2;8,4].*ones(1,1,size(b,3)), "f31v", dimInput, dimOutput);
defNode3.Label = "n3";
Node3 = Interpolation(defNode3);


defNode4.Domain = Domain(2);
f41v = VertexFunction(@(b) b.^2, @(b) [2*b(1,1,:),zeros(1,1,size(b,3));zeros(1,1,size(b,3)),2*b(2,1,:)], "f41v", dimInput, dimOutput);
f42v = VertexFunction(@(b) b.^3, @(b)  [3*b(1,1,:).^2,zeros(1,1,size(b,3));zeros(1,1,size(b,3)),3*b(2,1,:).^2], "f42v", dimInput, dimOutput);
defNode4.Children = [f41v,f42v];
defNode4.Label = "n4";
defNode4.Penalization = Penalization("simp",2);
Node4 = Interpolation(defNode4);

% Finally, we define the root and the all tree

defNode0.Domain = Domain("tetra");
defNode0.Children = [Node1,Node2,Node3,Node4];
defNode0.Label = "root";
defNode0.Penalization = Penalization("simp",2);
interpHierarchicalVector = Interpolation(defNode0);

interpHierarchicalVector.plotTree();
figure()
interpHierarchicalVector.plot();
%% 2) Definition of the variables
% a) Coordinates in the polygon

n = 1e3; % number of evaluation points
x = interpHierarchicalVector.initializeVariable(n,"rand",1)
x = interpHierarchicalVector.projection(x);
figure()
interpHierarchicalVector.plot(x);
% b) Vector field
% Each point $x$ is associated to a vector field value $b$.

b = rand(2,1,n);
%% 3) Evaluation of the interpolation
% a) Interpolated value

tic
val = interpHierarchicalVector.eval(x,b); % value
tv1 = toc;
disp(strcat("Compute interpolated values in ",num2str(tv1*1000)," ms"))
% b) Derivative w. r. t. the scalar field

tic
dfda =  interpHierarchicalVector.evalda(x,b); % derivative w.r.t a
tda1 = toc;
disp(strcat("Compute interpolated a-derivative in ",num2str(tda1*1000)," ms"))
% Check the Taylor expansion

epsilon = logspace(-8,-2,10); % test for different epsilon
res = zeros(10,1,n);

for i=1:10
    pert = epsilon(i)*rand(size(b));
    valPerturbeda = interpHierarchicalVector.eval(x,b+pert);
    res(i,1,:) = vecnorm(valPerturbeda - (val + mult(dfda,pert)),2,1);
end
MaxRes = max(res,[],3); medRes = median(res,3);
loglog(epsilon,medRes,'-o',epsilon,MaxRes,'-o',epsilon,epsilon.^2,'k--');
legend("med res","max res","\epsilon^2","Location","northwest")
xlabel("\epsilon")
ylabel("Norm of Taylor remainder")
title("Taylor remainder value w.r.t a")
grid on
% c) Derivative w. r. t. the position in the interpolation domain

tic
x0=x;
dfdx =  interpHierarchicalVector.evaldx(x,b); % derivative w.r.t x
tdx1 = toc;
disp(strcat("Compute interpolated x-derivative in ",num2str(tdx1*1000)," ms"))
% Check the Taylor expansion
% The computation of derivative in 3D domain is ill-conditionned near the boundaries. 
% Therefore, the |epsProj| parameter sets the minimum distance between the projected 
% points and the boundary. When |epsProj=0| the projection is exact but the computation 
% of derivative is wrong, and if |epsProj| is big the computation of derivative 
% is precise but the projection is wrong. |epsProj=1e-5| is a good tradeoff.

epsProj = 1e-5;
defNode0.Domain = Domain("tetra", 1, epsProj);
interpHierarchicalVector = Interpolation(defNode0);

x = interpHierarchicalVector.projection(x0);
fCorrected = interpHierarchicalVector.eval(x, b);
dfdxCorrected = interpHierarchicalVector.evaldx(x, b);

epsilon = logspace(-8,-2,10); % test for different epsilon
allLabel = [interpHierarchicalVector.getAllNodes.Label];
res = zeros(10,length(allLabel),n);
resXCorrected = zeros(10,length(allLabel),n);

for i=1:10
    for j=1:length(allLabel)
        l=allLabel(j);
        pert = epsilon(i)*rand(size(x.(l)));
        xPert = x;
        xPert.(l) = x.(l) + pert;
        xPert = interpHierarchicalVector.projection(xPert);
        pert = xPert.(l) - x.(l);
        valPerturbedx = interpHierarchicalVector.eval(xPert,b);
        pert = permute(pert,[2,3,1]);
        res(i,j,:) = vecnorm(valPerturbedx - (val + mult(dfdx.(l),pert)),2,1);
        resXCorrected(i,j,:) = vecnorm(valPerturbedx - (fCorrected + mult(dfdxCorrected.(l),pert)),2,1);
    end
end
MaxRes = max(res,[],[2,3]); medRes = median(res,[2,3]);
maxResXCorrected = max(resXCorrected,[],[2,3]); medResXCorrected = median(resXCorrected,[2,3]);

loglog(epsilon,medRes,'-o',epsilon,MaxRes,'-o'); hold on;
loglog(epsilon,maxResXCorrected,'-o',epsilon,medResXCorrected,'-o',epsilon,epsilon.^2,'k--'); hold off

legend("med res","max res", "med res corr.", "max res corr.", "\epsilon^2","Location","bestoutside")
xlabel("\epsilon")
ylabel("Norm of Taylor remainder")
title("Taylor remainder value w.r.t x")
grid on
%% 4) Speed tip
% To compute several times the interpolation for different $a$ without changing 
% the position $x$ in the domain, one should first compute the shape functions 
% once for all.

b2 = rand(2,100,n);
% slow
tic
for i=1:100
    interpHierarchicalVector.eval(x,b2(:,i,:)); % value
    interpHierarchicalVector.evalda(x,b2(:,i,:)); % derivative w.r.t a
    interpHierarchicalVector.evaldx(x,b2(:,i,:)); % derivative w.r.t x
end
tslow = toc;
disp(strcat("Computation time : ",num2str(tslow)," s"))

% fast
tic
 % pre-computation of the shape functions
[w,dwdx] = interpHierarchicalVector.evalBasisFunction(x);
for i=1:100
    interpHierarchicalVector.eval(x,b2(:,i,:),w); % value
    interpHierarchicalVector.evalda(x,b2(:,i,:),w); % derivative w.r.t a
    interpHierarchicalVector.evaldx(x,b2(:,i,:),w,dwdx); % derivative w.r.t x
end
tfast = toc;
disp(strcat("Computation time : ",num2str(tfast)," s"))

File: examples/python/py/e_HierarchicalVectorInterp.py
# -*- coding: utf-8 -*-
"""
Hierarchical vector interpolation example
"""


import matplotlib.pyplot as plt
import numpy as np
from ipywidgets import interact, interactive, fixed, interact_manual
import ipywidgets as widgets
import sys, os
sys.path.append(os.path.relpath("../../src"))
import hiped as hp

#%% 1) Definition of the interpolation tree
# From the leaves to the root.

from hiped.utils import mult

dimInput = 2
dimOutput = 2

# node 1,1 (3 vertices)
f111s = hp.VertexFunction("f111s(u)", f = lambda u: u**2,  dfdu = lambda u : 2*u * np.eye(2)[:,:,None], dimInput = dimInput, dimOutput = dimOutput)
f112s = hp.VertexFunction("f112s(u)", lambda u: u**3, lambda u : 3*u**2 * np.eye(2)[:,:,None], dimInput, dimOutput)
f113s = hp.VertexFunction("f113s(u)",  lambda u : 0.1*u, lambda u : 0.1 * np.eye(2)[:,:,None] * np.ones(u.shape), dimInput, dimOutput)
node11 = hp.Interpolation(hp.Domain(3), children = [f111s,f112s,f113s], label = "n11",
                          penalization = hp.Penalization("simp", 2))

# node 1,2 (2 vertices)
f121s = hp.VertexFunction("f121s(u)", lambda u: np.ones(u.shape), lambda u:  np.zeros(u.shape) * np.zeros((dimInput, dimOutput,1)), dimInput, dimOutput)
f122s = hp.VertexFunction("f122s(u)", lambda u: mult(np.array([[1,2],[3,4]]),u), lambda u: np.array([[1,2],[3,4]]).reshape(2,2,1)*np.ones(u.shape), dimInput, dimOutput)
node12 = hp.Interpolation(hp.Domain(2), children = [f121s,f122s], label = "n12",
                          penalization = hp.Penalization("simp", 2)) 

# node 1 (2 vertices)
node1 = hp.Interpolation(hp.Domain(2), children = [node11,node12],
                          label = "n1", penalization = hp.Penalization("simp",2)) 

# node 2 (singleton for vizualization) ; can be replaced directly by a VertexFunction
f2s = hp.VertexFunction("f2s(u)", lambda u: 0.2*u**3, lambda u: 0.6*u**2 * np.eye(2)[:,:,None], dimInput, dimOutput)
node2 = hp.Interpolation(hp.Domain(1), children = [f2s], label = "n2") 

# node 3 (singleton for vizualization)
f3s = hp.VertexFunction("f3s(u)", lambda u: mult(np.array([[1,2],[8,4]]),u), lambda u: np.array([[1,2],[8,4]]).reshape(2,2,1)*np.ones(u.shape), dimInput, dimOutput)
node3 = hp.Interpolation(hp.Domain(1), children = [f3s], label = "n3") 

# node 4
f41s = hp.VertexFunction("f41s(u)", lambda u: u**2, lambda u : 2*u * np.eye(2)[:,:,None], dimInput, dimOutput)
f42s = hp.VertexFunction("f42s(u)", lambda u: 0.1*u**3,  lambda u: 0.3*u**2 * np.eye(2)[:,:,None], dimInput, dimOutput)
node4 = hp.Interpolation(hp.Domain(2), children = [f41s, f42s], label = "n4",
                          penalization = hp.Penalization("simp", 2))

# Root
penalroot = hp.Penalization("zhou", 2)
interpHierarchical = hp.Interpolation(hp.Domain("tetra"), children = [node1, node2, node3, node4],
                          label = "root", penalization = penalroot)

# Interpolation tree
plt.figure()
interpHierarchical.plotTree()

#%% 2) Definition of the variables (coordinates x and scalar field u)

n = 100
radius = 1 # change to 0.3 to observe the effects
 # initialization of the variables
x = interpHierarchical.setInitialVariable(n, typeInit = "rand", radius = radius)
x = interpHierarchical.projection(x) # projection onto the domain
u = np.random.rand(dimInput,1,n)

# 3D vizualization of the hierarchical domain
plt.figure()
interpHierarchical.plot(x)
plt.show()

#%% 3) Evaluation of the interpolation
#a) Interpolated values

import time
nRuns = 500
t0 = time.time()
for i in range(nRuns):
    f = interpHierarchical.eval(x, u) # evaluate the interpolation
tv1 = time.time() - t0
print(f"Compute interpolated values in {tv1*1000:.3f} ms ({u.shape[2]} values, {nRuns} runs)")

#%% b) Derivative with respect to the scalar field u

nRuns = 500
t0 = time.time()
for i in range(nRuns):
    dfdu =  interpHierarchical.evaldu(x, u) # evaluate the derivative of the interpolation w.r.t u
tv1 = time.time() - t0
print(f"Compute u-derivative in {tv1*1000:.3f} ms ({u.shape[2]} values, {nRuns} runs)")

#%% Check the accuracy of the Taylor expansion

from hiped.utils import mult
h = np.logspace(-8,-2,10) # test for 10 different h
resU = np.zeros((10,1,u.shape[2]))


for i in range(10):
    pert = h[i]*np.random.rand(*u.shape);
    fPerturbedu = interpHierarchical.eval(x,u+pert);
    resU[i,0,:] = np.linalg.norm(fPerturbedu - (f + mult(dfdu,pert)), axis = 0)
maxResU = np.max(resU, axis = 2); medResU = np.median(resU, axis = 2)

plt.figure()
plt.loglog(h, medResU,'-o', label =  "median of the residual")
plt.loglog(h, maxResU,'-o', label =  "maximum of the residual")
plt.loglog(h, h**2,'k--', label =  "expected decay $(h^2)$")
plt.legend(loc = "best"); plt.grid()
plt.xlabel("$h$"); plt.ylabel("Euclidian norm of Taylor remainder")
plt.title("Taylor remainder with respect to the scalar field $u$")
plt.show()

#%% c) Derivative with respect to the cartesian coordinates x in the polygon

t0 = time.time()
for i in range(nRuns):
    dfdx =  interpHierarchical.evaldx(x, u) # evaluate the derivative of the interpolation w.r.t x
tv1 = time.time() - t0
print(f"Compute x-derivative in {tv1*1000:.3f} ms ({u.shape[2]} values, {nRuns} runs)")

#%% Check the accuracy of the Taylor expansion

'''
The relative error of dfdx is quite high in the 3D domain, due to the bad 
conditioning when the points lie near the boundaries. To avoid this phenomenon, 
the outside points are projected slightly inside the domain. The coefficient 
epsProj determines how deep. The deeper it is (high epsProj), the better is the 
computation of dfdx, but the worse is the projection. epsProj = 1e-5 (default 
value) is a good trade-off, but one can try other values.
The radius of the dispersion of the initial points  (section 2) can be reduced
to observe the difference on the Taylor tests.
'''

epsProj = 1e-5 # change to observe the effects

L = list(x.keys())
resX = np.zeros((10,len(L),u.shape[2]))
resXCorrected = np.zeros((10,len(L),u.shape[2]))

interpHierarchical = hp.Interpolation(hp.Domain("tetra", epsProj = epsProj),
                                      children = [node1, node2, node3, node4],
                                      label = "root", penalization = penalroot)

x = interpHierarchical.projection(x.copy())
fCorrected = interpHierarchical.eval(x, u)
dfdxCorrected = interpHierarchical.evaldx(x, u)

for i in range(10):
    for j in range(len(L)):
        xPert = x.copy()
        l=L[j]
        pert = h[i]*np.ones(x[l].shape)
        xPert[l] = x[l] + pert
        xPert = interpHierarchical.projection(xPert)
        pert = xPert[l] - x[l]
        valPerturbedx = interpHierarchical.eval(xPert,u)
        pert = pert[:,:,None].transpose(1,2,0)
        resX[i,j,:] = np.linalg.norm(valPerturbedx - (f + mult(dfdx[l], pert)), axis = 0)
        resXCorrected[i,j,:] = np.linalg.norm(valPerturbedx - (fCorrected + mult(dfdxCorrected[l], pert)), axis = 0)

maxResX = np.max(resX, axis = (1,2)); medResX = np.median(resX, axis =  (1,2))
maxResXCorrected = np.max(resXCorrected, axis = (1,2)); medResXCorrected = np.median(resXCorrected, axis =  (1,2))

plt.figure()
plt.loglog(h, medResX,'-o', label =  "median of the residual")
plt.loglog(h, maxResX,'-o', label =  "maximum of the residual")
plt.loglog(h, medResXCorrected,'-o', label =  "median of corrected residual")
plt.loglog(h, maxResXCorrected,'-o', label =  "maximum of corrected residual")
plt.loglog(h, h**2,'k--', label =  "expected decay $(h^2)$")
plt.legend(loc = "best"); plt.grid()
plt.xlabel("$h$"); plt.ylabel("Euclidian norm of Taylor remainder")
plt.title("Taylor remainder with respect to $x$")
plt.show()

#%% d) Speed tip

'''
To compute several times the interpolation for different $u$  without changing 
the position x in the domain (for example, when solving a non-linear system 
in u), one can first compute the shape functions once for all. '''

u2 = np.random.rand(dimInput,nRuns,u.shape[2])

# naive
t0 = time.time()
for i in range(nRuns):
    interpHierarchical.eval(x, u2[:,i,:]) # value
    interpHierarchical.evaldu(x, u2[:,i,:]) # derivative w.r.t u
    interpHierarchical.evaldx(x, u2[:,i,:]) # derivative w.r.t x
tNaive = time.time() - t0
print(f"Computation time (naive version, {nRuns} runs) : {tNaive*1000:.3f} ms")

# faster
t0 = time.time()
# pre-computation of the shape functions that don't depend on u
w, dwdx = interpHierarchical.evalBasisFunction(x)
for i in range(nRuns):
    interpHierarchical.eval(x, u2[:,i,:], w) # value
    interpHierarchical.evaldu(x, u2[:,i,:], w) # derivative w.r.t u
    interpHierarchical.evaldx(x, u2[:,i,:], w, dwdx) # derivative w.r.t x
tOptim = time.time() - t0
print(f"Computation time (optimized version, {nRuns} runs) : {tOptim*1000:.3f} ms")
